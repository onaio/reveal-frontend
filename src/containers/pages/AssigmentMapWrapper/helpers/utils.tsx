import { Dictionary } from '@onaio/utils';
import { Feature } from '@turf/turf';
import { History } from 'history';
import { EventData, Map } from 'mapbox-gl';
import { AssignmentMapWrapperProps } from '..';
import { PLAN_TYPES_WITH_MULTI_JURISDICTIONS } from '../../../../configs/env';
import { getPlanType } from '../../../../helpers/utils';
import { SELECTION_REASON } from '../../../../store/ducks/opensrp/hierarchies/constants';
import { TreeNode } from '../../../../store/ducks/opensrp/hierarchies/types';
import { nodeIsSelected } from '../../../../store/ducks/opensrp/hierarchies/utils';

/**
 * Handler to get current parent id from clicked jurisdiction on map
 * @param props - AssignmentMapWrapper component props
 * @param setMapParent - function to set map parent id to state
 * @param history - history session object
 */

export const onJurisdictionClick = (
  props: AssignmentMapWrapperProps,
  setMapParent: (currentId: string) => void,
  history: History
) => {
  function handleJurisdictionClick(_: Map, e: EventData) {
    // Destructure event data
    const { point, target, originalEvent } = e;
    const {
      currentParentNode,
      rootJurisdictionId,
      selectNodeCreator,
      deselectNodeCreator,
      currentChildren,
      plan,
      baseAssignmentURL,
    } = props;
    // grab underlying features from map
    const features: Feature[] = target.queryRenderedFeatures(point);
    if (!features.length) {
      return false;
    }

    for (const feature of features) {
      const activeJurisdictionId =
        (feature && feature.id) ||
        (feature && feature.properties && feature.properties.jurisdiction_id);
      if (activeJurisdictionId) {
        const currentId = activeJurisdictionId;
        let activeCurrentNode: TreeNode | undefined;

        if (currentParentNode) {
          // ensure currentparentnode is not undefined first
          // Handle selection for admin level 0
          if (
            currentParentNode.model.id === rootJurisdictionId &&
            currentId === currentParentNode.model.id
          ) {
            activeCurrentNode = currentParentNode;
          } else {
            // Handle selection for admin level > 0
            activeCurrentNode = currentChildren.find(
              (node: TreeNode) => node.model.id === currentId
            );
          }
        }
        const thePlanType = getPlanType(plan);
        if (
          activeCurrentNode &&
          originalEvent.altKey &&
          PLAN_TYPES_WITH_MULTI_JURISDICTIONS.includes(thePlanType)
        ) {
          if (!nodeIsSelected(activeCurrentNode)) {
            selectNodeCreator(
              rootJurisdictionId,
              activeCurrentNode.model.id,
              plan.identifier,
              SELECTION_REASON.USER_CHANGE
            );
          } else {
            deselectNodeCreator(
              rootJurisdictionId,
              activeCurrentNode.model.id,
              plan.identifier,
              SELECTION_REASON.USER_CHANGE
            );
          }
          setMapParent(currentId);
        } else {
          if (activeCurrentNode && activeCurrentNode.hasChildren()) {
            history.push(`${baseAssignmentURL}/${currentId}`);
          }
        }
        break;
      }
    }
  }
  return handleJurisdictionClick;
};

/**
 * Map component popup handler
 * @param map - mapbox map instance
 * @param event - map event object
 */

export const buildMouseMoveHandler = (map: Map, event: EventData) => {
  const popup = new mapboxgl.Popup({
    closeButton: false,
  });
  // Added this part to remove duplicate popup containers showing on map when hovering
  const nodes: any = document.getElementsByClassName('mapboxgl-popup');
  let node: HTMLDivElement;
  if (nodes && nodes.length) {
    for (node of nodes) {
      if (node.style) {
        node.style.display = 'none';
      }
    }
  }
  let content: string = '';
  let feature: Feature | any = {};
  popup.remove();
  // grab underlying features from map
  const features: any = event.target.queryRenderedFeatures(event.point) as Dictionary[];
  //
  if (!features.length) {
    return;
  }

  // Change the cursor style to pointer if features exist.
  map.getCanvas().style.cursor = features.length ? 'pointer' : 'grab';

  // loop through features, setting the name property to content var
  for (feature of features) {
    if (feature && feature.properties && feature.properties.name) {
      content += `<div class="jurisdiction-name"><center>${feature.properties.name}</center></div>`;
    }
    /**
     * Break out of loop once content is set
     * This helps in a case where we could have 2 or more identical features
     */
    if (content) {
      break;
    }
  }
  /**
   * Add popup to map after content is set
   */
  if (content.length) {
    popup
      .setLngLat(map.unproject(event.point))
      .setHTML(content)
      .addTo(map);
  } else {
    // clear existing popups if content is empty
    popup.remove();
    return;
  }
  return true;
};
