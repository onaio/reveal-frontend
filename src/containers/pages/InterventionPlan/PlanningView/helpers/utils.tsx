import { DrillDownColumn } from '@onaio/drill-down-table/dist/types';
import React from 'react';
import { Link } from 'react-router-dom';
import { Cell } from 'react-table';
import { DATE_CREATED, NAME, STATUS_HEADER } from '../../../../../configs/lang';
import { planStatusDisplay } from '../../../../../configs/settings';
import { ASSIGN_JURISDICTIONS_URL, DRAFT_IRS_PLAN_URL } from '../../../../../constants';
import { PlanRecord } from '../../../../../store/ducks/plans';

/** Columns definition for IRS drafts page table */
export const commonColumns: Array<DrillDownColumn<PlanRecord>> = [
  {
    Header: DATE_CREATED,
    accessor: 'plan_date',
    maxWidth: 50,
    minWidth: 20,
  },
  {
    Header: STATUS_HEADER,
    accessor: (d: PlanRecord) => planStatusDisplay[d.plan_status] || d.plan_status,
    id: 'plan_status',
    maxWidth: 50,
    minWidth: 20,
  },
];

export const irsDraftPageColumns: Array<DrillDownColumn<PlanRecord>> = [
  {
    Cell: (cell: Cell<PlanRecord>) => {
      const original = cell.row.original;
      return (
        <div>
          <Link to={`${DRAFT_IRS_PLAN_URL}/${original.id || original.plan_id}`}>{cell.value}</Link>
        </div>
      );
    },
    Header: NAME,
    accessor: 'plan_title',
    minWidth: 200,
  },
  ...commonColumns,
];

export const draftPageColumns: Array<DrillDownColumn<PlanRecord>> = [
  {
    Cell: (cell: Cell<PlanRecord>) => {
      const original = cell.row.original;
      return (
        <div>
          <Link to={`${ASSIGN_JURISDICTIONS_URL}/${original.id || original.plan_id}`}>
            {cell.value}
          </Link>
        </div>
      );
    },
    Header: NAME,
    accessor: 'plan_title',
    minWidth: 200,
  },
  ...commonColumns,
];
