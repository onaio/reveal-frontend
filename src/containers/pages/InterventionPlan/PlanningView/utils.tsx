import { DrillDownColumn } from '@onaio/drill-down-table/dist/types';
import React from 'react';
import { Link } from 'react-router-dom';
import { Cell } from 'react-table';
import { DATE_CREATED, NAME, STATUS_HEADER } from '../../../../configs/lang';
import { planStatusDisplay } from '../../../../configs/settings';
import { DRAFT_IRS_PLAN_URL } from '../../../../constants';
import { PlanRecord } from '../../../../store/ducks/plans';

/** Columns definition for IRS drafts page table */
export const columns: Array<DrillDownColumn<PlanRecord>> = [
  {
    Cell: (cell: Cell<PlanRecord>) => {
      const original = cell.row.original;
      return (
        <div>
          <Link to={`${DRAFT_IRS_PLAN_URL}/${original.id || original.plan_id}`}>{cell.value}</Link>
        </div>
      );
    },
    Header: NAME,
    accessor: 'plan_title',
    minWidth: 200,
  },
  {
    Header: DATE_CREATED,
    accessor: 'plan_date',
    maxWidth: 50,
    minWidth: 20,
  },
  {
    Header: STATUS_HEADER,
    accessor: (d: PlanRecord) => planStatusDisplay[d.plan_status] || d.plan_status,
    id: 'plan_status',
    maxWidth: 50,
    minWidth: 20,
  },
];
