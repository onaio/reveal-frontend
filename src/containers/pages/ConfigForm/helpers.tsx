import React from 'react';
import Loading from '../../../components/page/Loading';
import { NoDataComponent } from '../../../components/Table/NoDataComponent';
import { OPENSRP_API_BASE_URL } from '../../../configs/env';
import { getAcessTokenOrRedirect, growl } from '../../../helpers/utils';
import { newGetPayloadOptions } from '../../../services/opensrp';

/** drill down props */
export const drillDownProps = {
  paginate: false,
  renderNullDataComponent: () => <NoDataComponent />,
};

export const defaultConfigProps = {
  LoadingComponent: <Loading />,
  accessToken: getAcessTokenOrRedirect,
  baseURL: OPENSRP_API_BASE_URL,
  customAlert: growl,
  getPayload: newGetPayloadOptions,
};
