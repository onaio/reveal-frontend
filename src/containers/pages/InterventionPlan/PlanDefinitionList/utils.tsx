import { DrillDownColumn } from '@onaio/drill-down-table';
import { Dictionary } from '@onaio/utils/dist/types/types';
import React from 'react';
import { Link } from 'react-router-dom';
import { Cell, Row } from 'react-table';
import { isFIOrDynamicFI } from '../../../../components/forms/PlanForm/helpers';
import { DISPLAYED_PLAN_TYPES } from '../../../../configs/env';
import {
  CASE_TRIGGERED_TITLE,
  FOCUS_INVESTIGATION_STATUS_REASON,
  INTERVENTION_TYPE_LABEL,
  LAST_MODIFIED,
  ROUTINE_TITLE,
  STATUS_HEADER,
  TITLE,
} from '../../../../configs/lang';
import { planStatusDisplay, UseContext } from '../../../../configs/settings';
import {
  CASE_TRIGGERED,
  FI_REASON_CODE,
  INTERVENTION_TYPE_CODE,
  PLAN_UPDATE_URL,
  ROUTINE,
} from '../../../../constants';
import { InterventionType, PlanRecord } from '../../../../store/ducks/plans';

/**
 * sort plan UseContext values
 * @param {Row<PlanRecord>} rowA - table props
 * @param {Row<PlanRecord>} rowB - table props
 * @param {string} code - Plan use context code
 */
const sortUseContext = (rowA: Row<PlanRecord>, rowB: Row<PlanRecord>, code: string) => {
  /** custom sort function based useContext */
  const getIntervention = (row: Row<PlanRecord>) => {
    const interventionPlanContext = (row.original as Dictionary).plan_useContext.filter(
      (context: UseContext) => context.code === code
    );
    return interventionPlanContext.length > 0 ? interventionPlanContext[0].valueCodableConcept : '';
  };
  return getIntervention(rowA) > getIntervention(rowB) ? 1 : -1;
};

/**
 *
 * @param {UseContext[]} value - List of plan use contexts
 * @param {string} code - Plan use context code
 */
const getUseContext = (value: UseContext[], code: string) => {
  const typeUseContext = value.filter((e: any) => e.code === code);
  if (typeUseContext[0]?.valueCodableConcept === CASE_TRIGGERED) {
    return CASE_TRIGGERED_TITLE;
  }
  if (typeUseContext[0]?.valueCodableConcept === ROUTINE) {
    return ROUTINE_TITLE;
  }
  return typeUseContext.length > 0 ? typeUseContext[0].valueCodableConcept : null;
};

// plan intervention type column
const interventionTypeColumn = {
  Cell: ({ value }: Cell<PlanRecord>) => getUseContext(value, INTERVENTION_TYPE_CODE),
  Header: INTERVENTION_TYPE_LABEL,
  accessor: 'plan_useContext',
  id: 'plan_use_context_intervention_type',
  maxWidth: 40,
  minWidth: 10,
  sortType: (rowA: Row<PlanRecord>, rowB: Row<PlanRecord>) =>
    sortUseContext(rowA, rowB, INTERVENTION_TYPE_CODE),
} as DrillDownColumn<PlanRecord>;

// plan FI reason column
const fiReasonColumns = {
  Cell: ({ value }: Cell<PlanRecord>) => getUseContext(value, FI_REASON_CODE),
  Header: FOCUS_INVESTIGATION_STATUS_REASON,
  accessor: 'plan_useContext',
  id: 'plan_use_context_fi_reason',
  maxWidth: 40,
  minWidth: 10,
  sortType: (rowA: Row<PlanRecord>, rowB: Row<PlanRecord>) =>
    sortUseContext(rowA, rowB, FI_REASON_CODE),
} as DrillDownColumn<PlanRecord>;

const displayedPlans = [...DISPLAYED_PLAN_TYPES] as InterventionType[];
const showFiReasonColumns =
  displayedPlans.length === 2 &&
  isFIOrDynamicFI(displayedPlans[0]) &&
  isFIOrDynamicFI(displayedPlans[1]);

const interventionOrFiReasonColumn = showFiReasonColumns ? fiReasonColumns : interventionTypeColumn;

export const TableColumns: Array<DrillDownColumn<PlanRecord>> = [
  {
    Cell: (cell: Cell<PlanRecord>) => {
      const original = cell.row.original;
      return (
        <Link to={`${PLAN_UPDATE_URL}/${original.id}`} key={original.id}>
          {cell.value}
        </Link>
      );
    },
    Header: TITLE,
    accessor: 'plan_title',
    minWidth: 200,
  },
  interventionOrFiReasonColumn,
  {
    Cell: (cell: Cell<PlanRecord>) => {
      return planStatusDisplay[cell.value] || null;
    },
    Header: STATUS_HEADER,
    accessor: 'plan_status',
    maxWidth: 30,
    minWidth: 10,
  },
  {
    Header: LAST_MODIFIED,
    accessor: 'plan_date',
    maxWidth: 30,
    minWidth: 20,
  },
];
