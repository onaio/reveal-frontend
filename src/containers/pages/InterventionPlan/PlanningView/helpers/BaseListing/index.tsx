/**  underlying controlled component that fetches the plan data from
 * opensrp api and renders it.
 */
import { DrillDownTable, DrillDownTableProps } from '@onaio/drill-down-table';
import React from 'react';
import { NoDataComponent } from '../../../../../../components/Table/NoDataComponent';
import { PlanRecord } from '../../../../../../store/ducks/plans';
import { draftPageColumns } from '../utils';

/** table prop type */
export type BaseListTableProps = Pick<
  DrillDownTableProps<PlanRecord>,
  | 'columns'
  | 'data'
  | 'loading'
  | 'loadingComponent'
  | 'renderInBottomFilterBar'
  | 'renderInTopFilterBar'
  | 'useDrillDown'
  | 'renderNullDataComponent'
>;

/** gets props to be passed to drillDown Table */
export const getDefaultTableProps = (loading: boolean, data: PlanRecord[]) => {
  return {
    columns: draftPageColumns,
    data,
    loading,
    renderNullDataComponent: () => <NoDataComponent />,
    useDrillDown: false,
  };
};

/** props for base list component
 * responsibilities for this is to fetch data and render it in the table
 * it should receive a data loading function that tells it how to send the
 * data to the store after being saved.
 */
export interface BaseListComponentProps<TProps = BaseListTableProps> {
  loadData?: (setLoading: React.Dispatch<React.SetStateAction<boolean>>) => void;
  plansArray: PlanRecord[];
  getTableProps?: (loading: boolean, data: PlanRecord[]) => TProps;
}

const defaultBaseListProps = {
  plansArray: [],
};

/** pull data from the api and render table */
function BaseListComponent(props: BaseListComponentProps) {
  const [loading, setLoading] = React.useState<boolean>(props.plansArray.length === 0);

  const { loadData, getTableProps } = props;
  React.useEffect(() => {
    if (!loadData) {
      setLoading(false);
    } else {
      loadData(setLoading);
    }
  }, []);

  const tablePropsGetter = getTableProps ? getTableProps : getDefaultTableProps;

  const tableProps = tablePropsGetter(loading, props.plansArray);

  return <DrillDownTable {...tableProps} />;
}

BaseListComponent.defaultProps = defaultBaseListProps;

export { BaseListComponent };
