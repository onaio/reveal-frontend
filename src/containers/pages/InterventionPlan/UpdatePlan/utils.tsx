import { Dictionary } from '@onaio/utils/dist/types/types';
import {
  COMPLETE_PLAN_MESSAGE,
  FAILED_TO_GET_EVENT_ID,
  PLAN_CHANGES_HAVE_NOT_BEEN_SAVED,
  RETIRE_PLAN_MESSAGE,
} from '../../../../configs/lang';
import { PlanDefinition, UseContext } from '../../../../configs/settings';
import { displayError } from '../../../../helpers/errors';
import { infoGrowl } from '../../../../helpers/utils';
import { PlanStatus } from '../../../../store/ducks/plans';

/** util function to get the event id for case triggered plans
 * @params {PlanDefinition | null} aPlan - the plan to get the eventId from
 * @returns {string| null} The event id if present, null otherwise
 */
export const getEventId = (aPlan: PlanDefinition | null): string | null => {
  if (!aPlan) {
    return null;
  }
  const contextOfInterest = aPlan.useContext.filter((context: UseContext) => {
    return context.code === 'opensrpEventId';
  });

  if (contextOfInterest.length !== 1) {
    displayError(new Error(FAILED_TO_GET_EVENT_ID));
    return null;
  }
  return contextOfInterest[0].valueCodableConcept;
};

/** Util function to check if a plan was case Triggered or is just routine
 * @params {PlanDefinition} aPlan - the plan under inspection
 * @returns {boolean} - if the plan is reactive
 */
export const planIsReactive = (aPlan: PlanDefinition | null): boolean => {
  if (aPlan === null) {
    return false;
  }
  const contexts = aPlan.useContext;
  for (const context of contexts) {
    if (context.code === 'fiReason' && context.valueCodableConcept === 'Case Triggered') {
      return true;
    }
  }

  return false;
};

/** factory function for planform's beforeSubmit callback
 * @param originalPlan - plan before modifications
 * @returns {BeforeSubmit} callback function
 */
export const beforeSubmitFactory = (originalPlan: PlanDefinition) => (payload: PlanDefinition) => {
  const planStatusChanged = originalPlan.status !== payload.status;
  const interestingStatusChange =
    planStatusChanged &&
    [PlanStatus.RETIRED, PlanStatus.COMPLETE].includes(payload.status as PlanStatus);
  const messageLookup: Dictionary<string> = {
    [PlanStatus.COMPLETE]: COMPLETE_PLAN_MESSAGE,
    [PlanStatus.RETIRED]: RETIRE_PLAN_MESSAGE,
  };

  if (interestingStatusChange) {
    const res = window.confirm(messageLookup[payload.status]);
    if (!res) {
      infoGrowl(PLAN_CHANGES_HAVE_NOT_BEEN_SAVED);
    }
    return res;
  }
  return true;
};
