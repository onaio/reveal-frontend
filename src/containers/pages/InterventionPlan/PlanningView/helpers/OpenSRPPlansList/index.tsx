import { DrillDownColumn } from '@onaio/drill-down-table';
import reducerRegistry, { Registry } from '@onaio/redux-reducer-registry';
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { RouteComponentProps } from 'react-router';
import { ActionCreator, Store } from 'redux';
import Loading from '../../../../../../components/page/Loading';
import {
  defaultOptions,
  renderInFilterFactory,
} from '../../../../../../components/Table/DrillDownFilters/utils';
import { NoDataComponent } from '../../../../../../components/Table/NoDataComponent';
import {
  PLAN_RECORD_BY_ID,
  QUERY_PARAM_TITLE,
  QUERY_PARAM_USER,
} from '../../../../../../constants';
import {
  loadOpenSRPPlans,
  loadPlansByUserFilter,
} from '../../../../../../helpers/dataLoading/plans';
import { getQueryParams } from '../../../../../../helpers/utils';
import { OpenSRPService } from '../../../../../../services/opensrp';

import { DISPLAYED_PLAN_TYPES } from '../../../../../../configs/env';
import {
  LOADING,
  NO_DATA_FOUND,
  USER_HAS_NO_PLAN_ASSIGNMENTS,
} from '../../../../../../configs/lang';
import { displayError } from '../../../../../../helpers/errors';
import plansByUserReducer, {
  makePlansByUserNamesSelector,
  reducerName as plansByUserReducerName,
} from '../../../../../../store/ducks/opensrp/planIdsByUser';
import plansReducer, {
  fetchPlanRecords,
  FetchPlanRecordsAction,
  InterventionType,
  makePlansArraySelector,
  PlanRecord,
  reducerName as plansReducerName,
} from '../../../../../../store/ducks/plans';
import { BaseListComponent, BaseListComponentProps, BaseListTableProps } from '../BaseListing';

/** register the plans reducer */
reducerRegistry.register(plansReducerName, plansReducer);
reducerRegistry.register(plansByUserReducerName, plansByUserReducer);

/** minimal type for a renderProp */
export type RenderProp = () => React.ReactNode;

/** props for opensrpPlansList view */
export interface OpenSRPPlanListViewProps
  extends Omit<BaseListComponentProps, 'loadData' | 'getTableProps'> {
  fetchPlanRecordsCreator: ActionCreator<FetchPlanRecordsAction>;
  interventionTypes?: InterventionType[];
  planStatuses: string[];
  serviceClass: typeof OpenSRPService;
  sortByDate: boolean;
  tableColumns: Array<DrillDownColumn<PlanRecord>>;
  renderBody: (renderProp: RenderProp) => React.ReactNode;
  userName?: string | null;
  userNameFilter: boolean;
  noDataMessage: string;
}

/** default body render - allows to receive a JSX(as a render prop) from controlling component
 * that dictates how OpenSRPPlansList or its connected variation will sit
 * on the page that the controlling component is rendering.
 */
const defaultBodyRenderer = (componentRender: RenderProp) => {
  return <>{componentRender()}</>;
};

export const defaultProps: OpenSRPPlanListViewProps = {
  fetchPlanRecordsCreator: fetchPlanRecords,
  noDataMessage: NO_DATA_FOUND,
  planStatuses: [],
  plansArray: [],
  renderBody: defaultBodyRenderer,
  serviceClass: OpenSRPService,
  sortByDate: false,
  tableColumns: [],
  userNameFilter: false,
};

/** container view that renders opensrp plans from store , adds reveal-domain specific
 * stuff like the ui and functionality of the filters to be added to the table component
 */
const OpenSRPPlansList = (props: OpenSRPPlanListViewProps & RouteComponentProps) => {
  const {
    fetchPlanRecordsCreator,
    serviceClass,
    tableColumns,
    userNameFilter,
    noDataMessage,
  } = props;
  const loadData = (setLoading: React.Dispatch<React.SetStateAction<boolean>>) =>
    loadOpenSRPPlans(serviceClass, fetchPlanRecordsCreator, setLoading);
  const [loadingState, setLoadingState] = React.useState(!!props.userName);

  useEffect(() => {
    if (props.userName) {
      setLoadingState(true);
      loadPlansByUserFilter(props.userName)
        .finally(() => setLoadingState(false))
        .catch(err => displayError(err));
    }
  }, [props.userName]);

  const finalNoDataMessage = loadingState ? LOADING : noDataMessage;

  /** topbar filters */
  let topFilterBarParams: any = {
    ...defaultOptions,
    componentProps: props,
    queryParam: QUERY_PARAM_TITLE,
    showFilters: false,
  };

  if (userNameFilter) {
    topFilterBarParams = {
      ...topFilterBarParams,
      serviceClass: props.serviceClass,
      showFilters: true,
    };
  }

  const getTableProps = (loading: boolean, data: PlanRecord[]): BaseListTableProps => {
    return {
      columns: tableColumns,
      data,
      loading,
      loadingComponent: Loading,
      renderInBottomFilterBar: renderInFilterFactory({
        showColumnHider: false,
        showFilters: false,
        showPagination: true,
        showRowHeightPicker: false,
        showSearch: false,
      }),
      renderInTopFilterBar: renderInFilterFactory({
        ...topFilterBarParams,
      }),
      renderNullDataComponent: () => <NoDataComponent message={finalNoDataMessage} />,
      useDrillDown: false,
    };
  };

  const connectedBaseViewProps = {
    ...props,
    getTableProps,
    loadData,
  };

  const renderConnectedOpenSRPList = () => {
    return <BaseListComponent {...connectedBaseViewProps} />;
  };

  return <>{props.renderBody(renderConnectedOpenSRPList)}</>;
};

OpenSRPPlansList.defaultProps = defaultProps;

export { OpenSRPPlansList };

/** describes props returned by mapStateToProps */
export type MapStateToProps = Pick<OpenSRPPlanListViewProps, 'plansArray' | 'noDataMessage'>;
/** describe mapDispatchToProps object */
export type MapDispatchToProps = Pick<OpenSRPPlanListViewProps, 'fetchPlanRecordsCreator'>;
interface UserName {
  userName?: string | null;
}

/** data selector interface */
interface DataSelectors {
  interventionType: InterventionType[];
  planIds?: string[] | null;
  statusList: string[];
  title: string;
}

const plansByUserNameSelector = makePlansByUserNamesSelector();

/** maps props  to state */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: RouteComponentProps & OpenSRPPlanListViewProps
): MapStateToProps & UserName => {
  const plansArraySelector = makePlansArraySelector(PLAN_RECORD_BY_ID);
  const title = getQueryParams(ownProps.location)[QUERY_PARAM_TITLE] as string;
  const { planStatuses, interventionTypes } = ownProps;
  const interventionType =
    interventionTypes && interventionTypes.length
      ? interventionTypes
      : (DISPLAYED_PLAN_TYPES as InterventionType[]);
  const dataSelectors: DataSelectors = {
    interventionType,
    statusList: planStatuses,
    title,
  };
  // useName selector
  const userName = getQueryParams(ownProps.location)[QUERY_PARAM_USER] as string;
  let noDataMessage = NO_DATA_FOUND;
  if (userName) {
    const planIds = plansByUserNameSelector(state, { userName });
    dataSelectors.planIds = planIds;
    if (planIds?.length === 0) {
      noDataMessage = USER_HAS_NO_PLAN_ASSIGNMENTS;
    }
  }

  const plansRecordsArray = plansArraySelector(state as Registry, dataSelectors);
  // sort by date
  if (ownProps.sortByDate) {
    plansRecordsArray.sort(
      (a: PlanRecord, b: PlanRecord) => Date.parse(b.plan_date) - Date.parse(a.plan_date)
    );
  }
  const props = {
    noDataMessage,
    plansArray: plansRecordsArray,
    userName,
  };
  return props;
};

/** map action creators  */
const mapDispatchToProps: MapDispatchToProps = {
  fetchPlanRecordsCreator: fetchPlanRecords,
};

/** container creator that allows for configurable containers
 * the arguments passed here should respect the props that OpenSRPPlans list component take
 */
export const createConnectedOpenSRPPlansList = (
  mapState = mapStateToProps,
  mapDispatch = mapDispatchToProps
) => {
  return connect(mapState, mapDispatch)(OpenSRPPlansList);
};
