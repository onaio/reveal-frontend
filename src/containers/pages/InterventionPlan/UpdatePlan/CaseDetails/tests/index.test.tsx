import { mount, shallow } from 'enzyme';
import flushPromises from 'flush-promises';
import React from 'react';
import { Provider } from 'react-redux';
import ConnectedCaseDetails, { CaseDetails } from '..';
import { FAILED_TO_GET_EVENT_ID } from '../../../../../../configs/lang';
import * as utils from '../../../../../../helpers/errors';
import { OpenSRPService } from '../../../../../../services/opensrp';
import store from '../../../../../../store';
import { fetchEvents } from '../../../../../../store/ducks/opensrp/events';
import { rawEvent1 } from '../../../../../../store/ducks/opensrp/events/tests/fixtures';
import { extractEvent } from '../../../../../../store/ducks/opensrp/events/utils';

// tslint:disable-next-line: no-var-requires
const fetch = require('jest-fetch-mock');

const globalProps = {
  event: null,
  eventId: null,
  fetchEventsCreator: fetchEvents,
  service: OpenSRPService,
};

describe('src/containers/pages/interventionPlan/updateplan/caseDetials', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    fetch.resetMocks();
    jest.clearAllMocks();
  });

  it('renders without crashing', () => {
    shallow(<CaseDetails {...globalProps} />);
  });

  it('renders correctly', async () => {
    const props = {
      ...globalProps,
      event: extractEvent(rawEvent1),
      eventId: '',
    };

    // still need to mock out service.list request
    fetch.once(JSON.stringify({}));
    const wrapper = mount(<CaseDetails {...props} />);

    // resolve promise to get event
    await flushPromises();
    await new Promise<unknown>(resolve => setImmediate(resolve));

    // rendered text
    expect(wrapper.find('Ripple').length).toEqual(0);
    expect(wrapper.text()).toMatchSnapshot(`Case Details`);
  });

  it('renders nothing when eventId is null', async () => {
    const props = {
      ...globalProps,
      event: extractEvent(rawEvent1),
    };

    const displayErrorMock: jest.Mock = jest.fn();
    (utils as any).displayError = displayErrorMock;

    // still need to mock out service.list request
    fetch.once(JSON.stringify({}));
    const wrapper = mount(<CaseDetails {...props} />);

    // resolve promise to get event
    await flushPromises();
    await new Promise<unknown>(resolve => setImmediate(resolve));

    // rendered text
    expect(wrapper.find('Ripple').length).toEqual(0);
    expect(wrapper.text()).toMatchInlineSnapshot(`""`);

    // expect display error was called with correct message
    expect(displayErrorMock).toHaveBeenCalledWith(new Error(FAILED_TO_GET_EVENT_ID));
  });

  it('works well with store', async () => {
    // still need to mock out service.list request
    fetch.once(JSON.stringify(rawEvent1));
    const props = {
      ...globalProps,
      eventId: rawEvent1._id,
    };
    const wrapper = mount(
      <Provider store={store}>
        <ConnectedCaseDetails {...props} />
      </Provider>
    );

    // at this point event is null, thus should be loading
    expect(wrapper.find('Ripple').length).toEqual(1);

    // resolve promise to get event
    await flushPromises();
    await new Promise<unknown>(resolve => setImmediate(resolve));
    wrapper.update();

    // rendered text
    expect(wrapper.find('Ripple').length).toEqual(0);
    expect(wrapper.text()).toMatchSnapshot(`Case Details`);
  });

  it('Makes the correct api calls', async () => {
    fetch.once(JSON.stringify(rawEvent1));
    const props = {
      ...globalProps,
      eventId: rawEvent1.baseEntityId,
    };
    mount(
      <Provider store={store}>
        <ConnectedCaseDetails {...props} />
      </Provider>
    );

    await flushPromises();

    expect(fetch.mock.calls[0]).toEqual([
      'https://reveal-stage.smartregister.org/opensrp/rest/event/findById?id=88684506-605d-41db-b904-efbaf9795d2a',
      {
        headers: {
          accept: 'application/json',
          authorization: 'Bearer null',
          'content-type': 'application/json;charset=UTF-8',
        },
        method: 'GET',
      },
    ]);
  });
});
