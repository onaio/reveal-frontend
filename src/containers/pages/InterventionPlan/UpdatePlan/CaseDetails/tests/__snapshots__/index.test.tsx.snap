// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/containers/pages/interventionPlan/updateplan/caseDetials renders correctly: Case Details 1`] = `"Case Number: 191316000004593200125140915723Case DetailsCase InformationAge13Case Class.ADate of DiagnosisJanuary 25, 2020 12:00 AMHouse NumberInvestigation DateJanuary 24, 2020 12:00 AMCase Notif. DateJanuary 25, 2020 12:00 AMPatient NameชาวาดีSpeciesVSurnameดารากมล Foci of ResidenceClassificationA1Identifier88684506-605d-41db-b904-efbaf9795d2aNameแม่หิต (6305010804)"`;

exports[`src/containers/pages/interventionPlan/updateplan/caseDetials works well with store: Case Details 1`] = `"Case Number: 191316000004593200125140915723Case DetailsCase InformationAge13Case Class.ADate of DiagnosisJanuary 25, 2020 12:00 AMHouse NumberInvestigation DateJanuary 24, 2020 12:00 AMCase Notif. DateJanuary 25, 2020 12:00 AMPatient NameชาวาดีSpeciesVSurnameดารากมล Foci of ResidenceClassificationA1Identifier88684506-605d-41db-b904-efbaf9795d2aNameแม่หิต (6305010804)"`;
